include/envoy/upstream/load_balancer.h

```diff
@@ -57,5 +57,56 @@ class LoadBalancer {
 
 typedef std::unique_ptr<LoadBalancer> LoadBalancerPtr;
 
+/**
+ * Factory for load balancers.
+ */
+class LoadBalancerFactory {
+public:
+  virtual ~LoadBalancerFactory() {}
+
+  /**
+   * @return LoadBalancerPtr a new load balancer.
+   */
+  virtual LoadBalancerPtr create() PURE;
+};
+
+typedef std::shared_ptr<LoadBalancerFactory> LoadBalancerFactorySharedPtr;
+
+/**
+ * A thread aware load balancer is a load balancer that is global to all workers on behalf of a
+ * cluster. These load balancers are harder to write so not every load balancer has to be one.
+ * If a load balancer is a thread aware load balancer, the following semantics are used:
+ * 1) A single instance is created on the main thread.
+ * 2) The shared factory is passed to all workers.
+ * 3) Every time there is a host set update on the main thread, all workers will create a new
+ *    worker local load balancer via the factory.
+ *
+ * The above semantics mean that any global state in the factory must be protected by appropriate
+ * locks. Additionally, the factory *must not* refer back to the owning thread aware load
```

Comment:
Yeah, the idea here is that locks are basically never contended and are only used to guard a shared_ptr. I also used RW locks. I tried to figure out how to do a fully lockless implementation using only TLS, but the lifetime issues are quite complex around CDS and cluster add/removes. I think it's probably possible to do fully lockless, but it didn't seem necessary. I can add a TODO to look at this in a follow up if you think it would be best.

Reply:
Sorry I just remembered exactly what the issue is. The issue is that to do this only using TLS we would need to have a new TLS function that allows us to create a new object on behalf of each worker *on the main thread*, and then fan them out individually. This is different from what we do in general today, which is to create an object and set it on all workers, assuming that object is constant. I didn't want to do that additional work, but I can add comments to the TODO.

Reply:
Yep, TODO is fine here.